### ISSUE TITLE
DPI-scaled game windows have the cursor rendered in the wrong position (offset) in output

### ISSUE THREAD

**Issue by bertiebaggio** [obsproject/obs-studio#5229]
### Operating System Info

Windows 10

### OBS Studio Version

27.0.1

### OBS Studio Log URL

N/A

### Expected Behavior

The cursor should be in the same position in output (and OBS preview) as it is in-game.

https://user-images.githubusercontent.com/7524620/131675225-e521723e-7f4e-4098-8acd-dd6a85769548.mp4

(cursor  in-game and in output starts in upper left and ends on green dot in lower-right)

### Current Behavior

Cursor in output & preview is in the incorrect position, offset by the amount of scaling selected.

https://user-images.githubusercontent.com/7524620/131675134-e577dc29-3ae2-47ae-af81-54dd011cffa8.mp4

(cursor in-game starts in upper left and ends on green dot in lower-right, but in output disappears offscreen)

### Steps to Reproduce

1. Enable Windows Scaling, eg:

![scalesetting](https://user-images.githubusercontent.com/7524620/131674273-c0b8d666-4c6c-4c45-aa11-4b1a2e9ae562.png)

2. Run game 
3. Add game capture source in OBS and hook game
4. Observe difference in mouse cursor in game vs preview window

### Anything else we should know?

The difference in expected and actual output is best seen by rendering both cursors:

https://user-images.githubusercontent.com/7524620/131675417-12ded503-a5b5-4432-925c-35d199fb3803.mp4

I am submitting this issue as at time of writing there are a whole lot of irrelevant search results pointing to old threads on forums, etc; and so I can reference it in PR for additional background info.

**Comment by bertiebaggio** [comment]
Not all games result in this behaviour. Please feel free to contribute to this list!

## Errant Cursors (Confirmed)

 - Europa Universalis IV (DXGI)
```
16:51:43.567: [game-capture: 'Game Capture'] capture initializing!
16:51:43.567: [game-capture: 'Game Capture'] map id: CaptureHook_Texture_3017162_1
16:51:43.567: [game-capture: 'Game Capture'] Starting capture
16:51:43.567: [game-capture: 'Game Capture'] shared texture capture successful
16:51:47.461: [game-capture: 'Game Capture'] Hooked IDXGISwapChain::Present
16:51:47.461: [game-capture: 'Game Capture'] Hooked IDXGISwapChain::ResizeBuffers
16:51:47.461: [game-capture: 'Game Capture'] Hooked IDXGISwapChain1::Present1
16:51:47.461: [game-capture: 'Game Capture'] Hooked IDXGISwapChain::Release
16:51:47.461: [game-capture: 'Game Capture'] Hooked DXGI
```
- Streets of Rogue (d3d11)
```
16:53:17.485: [game-capture: 'Game Capture'] attempting to hook process: StreetsOfRogue.exe
16:53:17.493: [game-capture: 'Game Capture'] existing hook found, signaling process: r5apex.exe
16:53:17.493: [game-capture: 'Game Capture'] map id: CaptureHook_HookInfo19176
16:53:17.506: [game-capture: 'Game Capture'] d3d11 shared texture capture successful
16:53:17.517: [game-capture: 'Game Capture'] capture initializing!
16:53:17.517: [game-capture: 'Game Capture'] map id: CaptureHook_Texture_7145656_1
16:53:17.517: [game-capture: 'Game Capture'] Starting capture
16:53:17.517: [game-capture: 'Game Capture'] shared texture capture successful
```
 - Crusader Kings 3 (d3d11)

## Working Cursors

- Overwatch [reported by RytoEX]
- Risk of Rain 2 (d3d11)
- Shenzen IO (d3d11)
- Mahjong Nagomi (d3d11)
- Barotrauma (d3d11)

## Anecdotal Errant Cursors

_(from reports elsewhere, treat with caution!)_

- Sai
- "SDL2/ImGUI-based applications"
- Darkest Dungeon [x3]
- Path of Exile
- Starcraft
- Peggle [?]

**Comment by bertiebaggio** [comment]
Some research over the last day or two.

## More Info

- The cursor offset problem affects games that have been explicitly scaled using the override DPI setting, ie:

  <img src="https://user-images.githubusercontent.com/7524620/132079090-37284cf8-4d44-4464-ab8d-0b3b1b8a7042.png" height="400" title="DPI override" />

 Whereas games that have not been scaled themselves, but which exist on a scaled monitor have the cursor rendered in the correct position by OBS.

- This issue more commonly affects windowed games.
   However games which have been scaled and then fullscreened still have the wrong cursor offset (eg Streets of Rogue).

- Some games' cursors are seemingly unaffected by scaling (eg Overwatch, Risk of Rain 1/2, barotrauma, Terraria, etc).
  As far as I can tell from testing this seems to be because **they render their own cursor** and so OBS cursor rendering is not used.
 (ie `data->visible == false`)

 This means those games would be unaffected by any changes to OBS' cursor rendering code!

- In spite of the game's code /manifest being written in a DPI-aware manner, enabling a DPI scaling override **causes the game windows to lose DPI awareness**
  ie they are [`DPI_AWARENESS_UNAWARE`](https://docs.microsoft.com/en-us/windows/win32/hidpi/dpi-awareness-context) or `DPI_AWARENESS_CONTEXT_UNAWARE_GDISCALED` (verified in both Task manager and using [`GetWindowDpiAwarenessContext`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getwindowdpiawarenesscontext)).
  This means any queries as to their DPI (scaling) via [`GetDpiForWindow()`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getdpiforwindow) to return 96 regardless of the actual value ([Windows lies about DPI](https://stackoverflow.com/a/36864741/4647265)).

## Current Bug

My understanding is the bug arises due to an interaction between the window size returned by the hook, and how the cursor info is obtained and translated.

The width and height for the game capture'd window presently return the unscaled value for the window dimensions. (In testing this is also the case for [`GetClientRect()`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getclientrect)).

However, [game cursor capture](https://github.com/obsproject/obs-studio/blob/master/plugins/win-capture/cursor-capture.c#L199) uses [`GetCursorInfo()`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getcursorinfo), which returns the [screen position](https://docs.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-cursorinfo) of the cursor.

When OBS [renders the game cursor, it uses `ClientToScreen()`](https://github.com/obsproject/obs-studio/blob/master/plugins/win-capture/game-capture.c#L1864) to figure out where the window is on the screen, which is then used to transform (offset) the rendered cursor to ostensibly the correct position.

However, because of the above, **the cursor position is obtained relative to the screen, but the output is relative to the unscaled size**, and so cursor position/movement is magnified by the scaling factor in rendered output.

**Comment by RytoEX** [comment]
As far as I can tell, this was fixed by #5235 for OBS Studio 27.2, so I'm going to close this. If this is still an issue, please feel free to document that and we can reopen this.
